### PR TITLE
chore: add cache control parser

### PR DESCRIPTION
**Pull request by Noroth**
## Checklist

- [ ] I have discussed my proposed changes in an issue and have received approval to proceed.
- [ ] I have followed the coding standards of the project.
- [ ] Tests or benchmarks have been added or updated.

## Open Source AI Manifesto

This project follows the principles of the [Open Source AI Manifesto](https://human-oss.dev). Please ensure your contribution aligns with its principles.

## Summary by CodeRabbit

- **New Features**
  - Added support for parsing HTTP `Cache-Control` response headers.
  - Recognizes `max-age`, `s-maxage`, `no-store`, `no-cache`, `public`, and `private` directives.
  - Supports multiple header values, quoted field lists, case-insensitive directives, duplicate directives, and unknown directives.
  - Validates malformed directives and safely handles excessively large cache durations.

- **Tests**
  - Added comprehensive coverage for valid, invalid, duplicate, malformed, and edge-case header inputs, including fuzz testing.

### CODE REVIEW

File: v2/pkg/engine/cache/cache_control.go [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+	"strconv"
+	"strings"
+	"time"
+)
+
+// FieldNames is a map of field names.
+type FieldNames struct {
+	// lockFields is a flag to prevent adding any fields.
+	// This is set when an empty field list is parsed.
+	lockFields bool
```
Comment (fiam) [inline v2/pkg/engine/cache/cache_control.go]: question: do we need this functionality? It doesn't seem particularly well done either, since `Add` doesn't give any feedback to the caller.

File: v2/pkg/engine/cache/cache_control.go [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+	// Present indicates whether the directive is present in the response.
+	Present bool
```
Comment (fiam) [inline v2/pkg/engine/cache/cache_control.go]: question: could this be done via returning a `nil` when field names are not present?

File: v2/pkg/engine/cache/cache_control.go [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+	// Fields is a list of field names.
+	// Do not add any fields to this map directly. Use the add method instead
+	Fields map[string]struct{}
```
Comment (fiam) [inline v2/pkg/engine/cache/cache_control.go]: nit: this might be better expressed as an API contract (i.e. not export `Fields`)

File: v2/pkg/engine/cache/cache_control.go [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+func NewFieldNames() *FieldNames {
+	return &FieldNames{
+		// By default, when a FieldNames is created, it is present.
+		Present: true,
+	}
+}
+
+func (f *FieldNames) Add(name string) {
+	if f.Fields == nil {
```
Comment (fiam) [inline v2/pkg/engine/cache/cache_control.go]: nit: assuming most of the time a Cache-Control header is present it will have non-zero fields, nitializing `.Fields` on `NewFieldNames` might be more efficient.

File: v2/pkg/engine/cache/cache_control.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+		if field == "" {
+			continue
+		}
+
+		fields.Add(field)
+		added = true
+	}
+
+	if !added {
+		fields.ClearAndLock()
```
Comment (fiam) [inline v2/pkg/engine/cache/cache_control.go]: question: isn't this redundant? if we haven't added any fields, what does clearing the inner map do?

File: v2/pkg/engine/cache/cache_control.go [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+	// whole response, not just a few fields. So an empty argument wins, and we
+	// lock the set so a later occurrence can't weaken it.
+	if !arg.present {
+		fields.ClearAndLock()
+		return
+	}
+
+	var added bool
+	for field := range strings.SplitSeq(arg.text, ",") {
+		field = strings.TrimSpace(field)
```
Comment (fiam) [inline v2/pkg/engine/cache/cache_control.go:286]: question: could the lexer/parser take care of this and absorb non-significant whitespace?

File: v2/pkg/engine/cache/cache_control.go [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+			return nil
+		}
+		seen.sMaxAge = true
+
+		if value, ok := deltaSecondsArgument(arg); ok {
+			cc.SMaxAge = &value
+		}
+
+	case "no-store":
+		// A stray argument means nothing to a boolean directive; discard it.
```
Comment (fiam) [inline v2/pkg/engine/cache/cache_control.go]: question: isn't this comment a bit confusing? it says `discard it` but then we set a variable to `true` just after it.

**Review by fiam — COMMENTED** [review]
not an exhaustive review since this is still in draft, just some comments. Thanks!

File: v2/pkg/engine/cache/cache_control.go (hunk 1 above)
Comment (Noroth) [reply v2/pkg/engine/cache/cache_control.go]: updated to `noFields`

File: v2/pkg/engine/cache/cache_control.go [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+
+	if num > math.MaxInt32 {
+		return DeltaSeconds(math.MaxInt32), nil
+	}
+
+	return DeltaSeconds(int32(num)), nil
+}
+
+// ToGoSeconds converts a DeltaSeconds to a time.Duration.
+func (d DeltaSeconds) ToGoSeconds() time.Duration {
```
Comment (fiam) [inline v2/pkg/engine/cache/cache_control.go]: nit: maybe just `Duration()` is more idiomatic?

File: v2/pkg/engine/cache/cache_control.go (hunk 2 above)
Comment (Noroth) [reply v2/pkg/engine/cache/cache_control.go]: Yeah we can do that

File: v2/pkg/engine/cache/cache_control.go (hunk 3 above)
Comment (Noroth) [reply v2/pkg/engine/cache/cache_control.go]: Made fields private and added an iterator function

File: v2/pkg/engine/cache/cache_control.go (hunk 4 above)
Comment (Noroth) [reply v2/pkg/engine/cache/cache_control.go]: done

**Comment by Noroth** [comment]
@coderabbitai review

**Review by SkArchon — COMMENTED** [review]
I think some verbs are missing like :"stale-while-revalidate / stale-if-error". Might be worth adding for correctness even though I don't think we will consider these in the first versions.

File: v2/pkg/engine/cache/cache_control.go [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+func ParseCacheControlResponse(headers http.Header) (*CacheControlResponse, error) {
+	// Nothing but whitespace and delimiters carries no directives, which is
+	// indistinguishable from the header being absent.
+	value := strings.Trim(strings.Join(headers.Values("Cache-Control"), ","), "\t\r\n")
+	if value == "" {
+		return &CacheControlResponse{}, nil
+	}
+
+	return parse(value)
+}
```
[the remark this answers is not in the document]
Comment (Noroth) [reply v2/pkg/engine/cache/cache_control.go:209]: I think it is okay in this scenario. We don't have to be too strict with the overall input string. We still check the individual elements within the input string for invalid characters. Trimming the input string is fine.

File: v2/pkg/engine/cache/cache_control.go (hunk 6 above)
Comment (Noroth) [reply v2/pkg/engine/cache/cache_control.go:286]: We could but I think this is okay for now. Moving the logic to the lexer would make it more complicated to read and we would have to emit more tokens and incorporate that in the parser logic.

**Review by SkArchon — APPROVED** [review]
lgtm

File: v2/pkg/engine/cache/cache_control.go [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+	if num > math.MaxInt32 {
+		return DeltaSeconds(math.MaxInt32), nil
+	}
+
+	return DeltaSeconds(int32(num)), nil
+}
+
+// AsDuration converts a DeltaSeconds to a time.Duration.
+func (d DeltaSeconds) AsDuration() time.Duration {
+	return time.Duration(d) * time.Second
```
[the remark this answers is not in the document]
Comment (Noroth) [reply v2/pkg/engine/cache/cache_control.go:123]: This PR adds the function, there are no consumers
